File: AurieSharpManaged/IAurie.cs
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace Aurie
{
    public enum AurieStatus : Int32
    {
        // The operation completed successfully.
        Success = 0,
        // An invalid architecture was specified.
        InvalidArch,
        // An error occured in an external function call.
        ExternalError,
        // The requested file was not found.
        FileNotFound,
        // The requested access to the object was denied.
        AccessDenied,
        // An object with the same identifier / priority is already registered.
        ObjectAlreadyExists,
        // One or more parameters were invalid.
        InvalidParameter,
        // Insufficient memory is available.
        InsufficientMemory,
        // An invalid signature was detected.
        InvalidSignature,
        // The requested operation is not implemented.
        NotImplemented,
        // An internal error occured in the module.
        ModuleInternalError,
        // The module failed to resolve dependencies.
        ModuleDependencyNotResolved,
        // The module failed to initialize.
        ModuleInitializedFailed,
        // The target file header, directory, or RVA could not be found or is invalid.
        FilePartNotFound,
        // The object was not found.
        ObjectNotFound,
        // The requested resource is unavailable.
        Unavailable
    }

    public class IAurie
    {
        private delegate IntPtr GetFrameworkRoutineDelegate([MarshalAs(UnmanagedType.LPStr)] string RoutineName);
        private GetFrameworkRoutineDelegate m_GetFrameworkRoutine;

        [MethodImpl(MethodImplOptions.NoInlining)]
        private string GetCurrentMethodName()
        {
            var st = new StackTrace();
            var sf = st.GetFrame(1);

            return sf?.GetMethod()?.Name ?? "";
        }

        private delegate AurieStatus ElIsProcessSuspendedDelegate(
            [MarshalAs(UnmanagedType.I1)] ref bool Suspended
        );
        public AurieStatus ElIsProcessSuspended(
            ref bool Suspended
        )
        {
            IntPtr function = m_GetFrameworkRoutine(GetCurrentMethodName());
            return Marshal.GetDelegateForFunctionPointer<ElIsProcessSuspendedDelegate>(function)(ref Suspended);
        }

        private delegate void MmGetFrameworkVersionDelegate(
            ref short Major,
            ref short Minor,
            ref short Patch
        );
        public void MmGetFrameworkVersion(
            ref short Major,
            ref short Minor,
            ref short Patch
        )
        {
            IntPtr function = m_GetFrameworkRoutine(GetCurrentMethodName());
            Marshal.GetDelegateForFunctionPointer<MmGetFrameworkVersionDelegate>(function)(
                ref Major,
                ref Minor,
                ref Patch
            );
        }
        private delegate IntPtr MmSigscanModuleDelegate(
            [MarshalAs(UnmanagedType.LPWStr)] string ModuleName,
            [MarshalAs(UnmanagedType.LPStr)] string Pattern,
            [MarshalAs(UnmanagedType.LPStr)] string PatternMask
        );

        public IntPtr MmSigscanModule(
            string ModuleName,
            string Pattern,
            string PatternMask
        )
        {
            IntPtr function = m_GetFrameworkRoutine(GetCurrentMethodName());
            return Marshal.GetDelegateForFunctionPointer<MmSigscanModuleDelegate>(function)(
                ModuleName,
                Pattern,
                PatternMask
            );
        }

        private delegate IntPtr MmSigscanRegionDelegate(
            UIntPtr RegionBase,
            nuint RegionSize,
            [MarshalAs(UnmanagedType.LPStr)] string Pattern,
            [MarshalAs(UnmanagedType.LPStr)] string PatternMask
        );

        public IntPtr MmSigscanRegion(
            UIntPtr RegionBase,
            nuint RegionSize,
            string Pattern,
            string PatternMask
        )
        {
            IntPtr function = m_GetFrameworkRoutine(GetCurrentMethodName());
            return Marshal.GetDelegateForFunctionPointer<MmSigscanRegionDelegate>(function)(
                RegionBase,
                RegionSize,
                Pattern,
                PatternMask
            );
        }

        // MmCreateHook omitted, no support for std::string_view
        // MmHookExists omitted, no support for std::string_view
        // MmGetHookTrampoline omitted, no support for std::string_view
        // MmRemoveHook omitted, no support for std::string_view
        // Md* omitted

        [return: MarshalAs(UnmanagedType.I1)]
        private delegate bool ObInterfaceExistsDelegate(
            [MarshalAs(UnmanagedType.LPStr)] string InterfaceName
        );

        public bool ObInterfaceExists(
            string InterfaceName
        )
        {
            IntPtr function = m_GetFrameworkRoutine(GetCurrentMethodName());
            return Marshal.GetDelegateForFunctionPointer<ObInterfaceExistsDelegate>(function)(
                InterfaceName
            );
        }

        private delegate AurieStatus ObGetInterfaceDelegate(
            [MarshalAs(UnmanagedType.LPStr)] string InterfaceName,
            ref IntPtr Interface
        );

        public AurieStatus ObGetInterface(
            string InterfaceName,
            ref IntPtr Interface
        )
        {
            IntPtr function = m_GetFrameworkRoutine(GetCurrentMethodName());
            return Marshal.GetDelegateForFunctionPointer<ObGetInterfaceDelegate>(function)(
                InterfaceName,
                ref Interface
            );
        }

        private delegate AurieStatus ObpLookupInterfaceOwnerExportDelegate(
            [MarshalAs(UnmanagedType.LPStr)] string InterfaceName,
            [MarshalAs(UnmanagedType.LPStr)] string ExportName,
            ref IntPtr ExportAddress
        );

        public AurieStatus ObpLookupInterfaceOwnerExport(
            string InterfaceName,
            string ExportName,
            ref IntPtr ExportAddress
        )
        {
            IntPtr function = m_GetFrameworkRoutine(GetCurrentMethodName());
            return Marshal.GetDelegateForFunctionPointer<ObpLookupInterfaceOwnerExportDelegate>(function)(
                InterfaceName,
                ExportName,
                ref ExportAddress
            );
        }

        public IAurie(IntPtr GetFrameworkRoutine)
        {
            m_GetFrameworkRoutine = Marshal.GetDelegateForFunctionPointer<GetFrameworkRoutineDelegate>(
                GetFrameworkRoutine
            );
        }
    }
}

File: AurieSharp/source/Managed/IAurie.cpp
#include "IAurie.hpp"
using Marshal = System::Runtime::InteropServices::Marshal;

namespace AurieSharp
{
	AurieStatus AurieInterface::ElIsProcessSuspended(
		OUT bool% Suspended
	)
	{
		printf("ElProcessSuspended called!!!");
		Beep(1000, 100);

		pin_ptr<bool> suspended = &Suspended;
		return static_cast<AurieStatus>(Aurie::ElIsProcessSuspended(*suspended));
	}

	void AurieInterface::MmGetFrameworkVersion(
		OUT short% Major,
		OUT short% Minor,
		OUT short% Patch
	)
	{
		pin_ptr<short> major = &Major;
		pin_ptr<short> minor = &Minor;
		pin_ptr<short> patch = &Patch;

		return Aurie::MmGetFrameworkVersion(
			major,
			minor,
			patch
		);
	}

	System::UIntPtr AurieInterface::MmSigscanModule(
		IN System::String^ ModuleName,
		IN System::String^ Pattern,
		IN System::String^ PatternMask
	)
	{
		// Allocates memory, we have to free it with FreeHGlobal after we're done
		System::IntPtr module_name = Marshal::StringToHGlobalUni(ModuleName);
		System::IntPtr pattern = Marshal::StringToHGlobalUni(Pattern);
		System::IntPtr pattern_mask = Marshal::StringToHGlobalAnsi(PatternMask);

		size_t result = Aurie::MmSigscanModule(
			reinterpret_cast<const wchar_t*>(module_name.ToPointer()),
			reinterpret_cast<const unsigned char*>(pattern.ToPointer()),
			reinterpret_cast<const char*>(pattern.ToPointer())
		);

		// Free all allocated memory
		Marshal::FreeHGlobal(module_name);
		Marshal::FreeHGlobal(pattern);
		Marshal::FreeHGlobal(pattern_mask);

		return System::UIntPtr(result);
	}

	System::UIntPtr AurieInterface::MmSigscanRegion(
		IN System::UIntPtr RegionBase,
		IN int RegionSize,
		IN System::String^ Pattern,
		IN System::String^ PatternMask
	)
	{
		// Allocates memory, we have to free it with FreeHGlobal after we're done
		System::IntPtr pattern = Marshal::StringToHGlobalUni(Pattern);
		System::IntPtr pattern_mask = Marshal::StringToHGlobalAnsi(PatternMask);

		size_t result = Aurie::MmSigscanRegion(
			reinterpret_cast<const unsigned char*>(RegionBase.ToPointer()),
			static_cast<size_t>(RegionSize),
			reinterpret_cast<const unsigned char*>(pattern.ToPointer()),
			reinterpret_cast<const char*>(pattern.ToPointer())
		);

		// Free all allocated memory
		Marshal::FreeHGlobal(pattern);
		Marshal::FreeHGlobal(pattern_mask);

		return System::UIntPtr(result);
	}
}

